File: gradientDescent/gradientDescent.go
package main

import (
	"encoding/csv"
	"fmt"
	"log"
	"math"
	"os"
	"strconv"

	"gonum.org/v1/gonum/stat"
)

func main() {
	f, err := os.Open("../data/Advertising_ISLR.csv")
	if err != nil {
		log.Fatal(err)
	}
	defer f.Close()

	reader := csv.NewReader(f)

	reader.FieldsPerRecord = 5
	trainingData, err := reader.ReadAll()
	if err != nil {
		log.Fatal(err)
	}

	x := make([][]float64, reader.FieldsPerRecord-1)
	var y []float64

	for f := 0; f < reader.FieldsPerRecord; f++ {
		for i, record := range trainingData {

			// Skip header
			if i == 0 {
				continue
			}

			if f == 0 {
				// Fill the bias with ones
				x[f] = append(x[f], 1)
			} else if f < reader.FieldsPerRecord-1 {
				// Populate Features
				xv, err := strconv.ParseFloat(record[f], 64)
				if err != nil {
					log.Fatal(err)
				}
				x[f] = append(x[f], xv)
			} else {
				// Populate Labels
				yv, err := strconv.ParseFloat(record[f], 64)
				if err != nil {
					log.Fatal(err)
				}
				y = append(y, yv)
			}
		}
	}

	// Standardize X to avoid under / over flow
	x = standardize(x)
	iterations := 1000
	learningRate := 0.1

	coefficients := train(x, y, iterations, learningRate)
	fmt.Println("TV, Radio, Newspaper features have been standardized")
	fmt.Println("Coefficients : ", coefficients)
}

func standardize(x [][]float64) [][]float64 {
	for i := 1; i < len(x); i++ {
		xmean := stat.Mean(x[i], nil)
		// Unbiased variance
		uxvariance := stat.Variance(x[i], nil)
		for j := 0; j < len(x[i]); j++ {
			// (x - mean) / std
			x[i][j] = (x[i][j] - xmean) / math.Pow(uxvariance, 0.5)
		}
	}
	return x
}

func train(x [][]float64, y []float64, iterations int, learningRate float64) []float64 {

	coefficients := make([]float64, len(x))

	for i := 0; i < iterations; i++ {
		gradientDescent(x, y, coefficients, learningRate)
	}

	return coefficients
}

func gradientDescent(x [][]float64, y []float64, coefficients []float64, learningRate float64) []float64 {

	residuals := make([]float64, len(y))
	for i := 0; i < len(y); i++ {
		prediction := 0.0
		for j := 0; j < len(coefficients); j++ {
			prediction += coefficients[j] * x[j][i]
		}
		residuals[i] = y[i] - prediction
	}

	for i := 0; i < len(residuals); i++ {
		// https://mccormickml.com/2014/03/04/gradient-descent-derivation/
		// Andrew Ng
		// We multiply our MSE cost function by 1/2 so that when we take the derivative, the 2s cancel out
		// Multiplying the cost function by a scalar does not affect the location of its minimum
		coefficients[0] += learningRate * residuals[i] / float64(len(residuals))
		for j := 1; j < len(x); j++ {
			coefficients[j] += learningRate * (residuals[i] * x[j][i]) / float64(len(residuals))
		}
	}

	return coefficients
}

File: gradientDescent/gradientDescent.js
const data = require("../data/advertising.json");
var _ = require('./lodash.js');

const x = [];
const y = [];
// const selectedFeatures = ['TV', 'Radio', 'Newspaper'];
const selectedFeatures = ['TV'];
const selectedLabels = ['Sales'];

Object.keys(data).forEach(function(key, keyIndex) {
    var temp = [];
    Object.keys(data[key])
      .filter(key => selectedFeatures.includes(key))
      .reduce((obj, kee) => {
        temp.push(data[key][kee]);
    }, {});
    x.push(temp);
});

Object.keys(data).forEach(function(key, keyIndex) {
    var temp = [];
    Object.keys(data[key])
      .filter(key => selectedLabels.includes(key))
      .reduce((obj, kee) => {
        temp.push(data[key][kee]);
    }, {});
    y.push(temp);
});

class GradientDescentJS {
    constructor(features, labels, options) {
        this.features = features;
        this.labels = labels;
        this.mseHistory = [];
        this.N = features.length;

        this.getStats();
    
        // assign default values to options
        this.options = Object.assign(
          { learningRate: 0.1, iterations: 10 },
          options
        );
        
        // standardize
        // uncomment yHat and mSlope
        this.features = this.features.map((val, i) => {
            return (val - this.xMean) / Math.pow(this.uxVariance, 0.5);
        });
    
        this.b = 0;
        this.m = 0;
    }

    gradientDescent() {
        const yHat = this.features.map(row => {
            // return this.m * row[0] + this.b
            return this.m * row + this.b
        });

        this.mse = _.sum(yHat.map((val, i) => {
            return Math.pow((this.labels[i][0] - val), 2) / (this.labels.length);
        }));

        this.mseHistory.unshift(this.mse);
        
        const bSlope = (_.sum(yHat.map((guess, i) => {
            return guess - this.labels[i][0]
        })) * 2) / this.features.length;

        const mSlope = (_.sum(yHat.map((guess, i) => {
            // return -1 * this.features[i][0] * (this.labels[i][0] - guess)
            return -1 * this.features[i] * (this.labels[i][0] - guess);
        })) * 2) / this.features.length;

        this.m = this.m - mSlope * this.options.learningRate;
        this.b = this.b - bSlope * this.options.learningRate;
    }

    train() {
        for( let i = 0; i < this.options.iterations; i++ ) {
            this.gradientDescent();
            this.updateLearningRate();
        }        
    }
    
    updateLearningRate() {
        if (this.mseHistory.length < 2) {
            return;
        }
    
        if (this.mseHistory[0] > this.mseHistory[1]) {
            this.options.learningRate /= 2;
        } else {
            this.options.learningRate *= 1.05;
        }
    }

    getStats() {
        // stats for x
        this.xMean = _.sum(this.features.map((val, i) => {
            return val / this.features.length;
        }));

        this.xVariance = _.sum(this.features.map((val, i) => {
            return Math.pow((val - this.xMean), 2) / this.features.length;
        }));

        this.uxVariance = _.sum(this.features.map((val, i) => {
            return Math.pow((val - this.xMean), 2) / (this.features.length - 1);
        }));

        // stats for y
        this.yMean = _.sum(this.labels.map((val, i) => {
            return val / this.labels.length;
        }));

        this.yVariance = _.sum(this.labels.map((val, i) => {
            return Math.pow((val - this.yMean), 2) / this.labels.length;
        }));

        this.uyVariance = _.sum(this.labels.map((val, i) => {
            return Math.pow((val - this.yMean), 2) / (this.labels.length - 1);
        }));
    }
}

regression = new GradientDescentJS(x, y, { learningRate: 0.1, iterations: 100 })
console.log("N: ", regression.N)
console.log("xMean: ", regression.xMean);
console.log("yMean: ", regression.yMean);
console.log("xVariance: ", regression.xVariance);
console.log("unbiased xVariance: ", regression.uxVariance);
console.log("yVariance: ", regression.yVariance);
console.log("unbiased yVariance: ", regression.uyVariance);
regression.train()
console.log("slope: ", regression.m)
console.log("bias: ", regression.b)